### ISSUE TITLE
Suggestion: Ugoira to GIF conversion

### ISSUE THREAD

**Issue by worldendDominator** [pseudonymous/better-better-booru#56]
Add an option go ugoira images to make a gif and download it. I know there are such scripts for Pixiv.

**Comment by Moebius-Strip** [comment]
Something like [Ugoira2GIF](https://greasyfork.org/en/scripts/6534-ugoira2gif) is still beyond me, but it can be easily modified to work with Danbooru. Try out this [modified version](https://gist.github.com/Moebius-Strip/3f819e65900445ba2ccc) I've posted to Gist. You can click "raw" or download it in as a ZIP to install it. When it's installed, the GIF download link will show up beside the video download link on ugoira posts.

It's a quick and dirty job, so let me know if something is acting up.

**Comment by worldendDominator** [comment]
I have tested it, and it breaks the same way as the original Pixiv script: freezing at 96%-98% of unzipping (line #990) when doing large images.
For comparison, it worked with a 1 MB 350x450 image (although still froze for a bit at 90%), but didn't work with a 5 MB 550x600 image a well as another larger one.
I cleared my cookies and it didn't help.

Also, a non-functional "Save as GIF" link also appears in posts with webm-converted ugoira (which have both tags).

**Comment by Moebius-Strip** [comment]
> Also, a non-functional "Save as GIF" link also appears in posts with webm-converted ugoira (which have both tags).

I actually caught and [fixed that](https://gist.github.com/Moebius-Strip/3f819e65900445ba2ccc/revisions#diff-55bdeec32f68d79f95ec1d4dea2537d7) about 20 minutes before your comment, but you must have downloaded it before then.

>I have tested it, and it breaks the same way as the original Pixiv script: freezing at 96%-98% of unzipping (line #990) when doing large images.

Depending on the limitations of the browser and/or computer you're using, it may or may not freeze/fail, but it will almost certainly take a long time. I was able to download the biggest ugoira post at 29.9 MB and turn it into a 59.9 MB GIF. Firefox did stay locked up most of the time though and also tended to eat through memory. It got even worse if the unresponsive script warning was disabled. If memory use got too high, it just failed with a generic error message in the console that described nothing.

One thing I noted was that Chrome handled this script much better than Firefox. When looking into it, it seems to be connected to the preparation of the frames which runs in a loop without pause. Chrome handles this loop differently than Firefox. It does it without locking up the browser for an extended period and allows memory to be cleaned out as things run. 

I've modified the [script](https://gist.github.com/Moebius-Strip/3f819e65900445ba2ccc) some more to allow Firefox to work during the preparation process and made the link display a percent status for the "preparation" stage. I've basically forced the loop to pause for 1/40 of a second after preparing each frame. This will prevent the unresponsive script warning while still allowing the tab to be closed/refreshed if things take too long. However, it does extend the total time by 1 second for every 40 frames and will cause Firefox to be very slow/jerky during the time when it previously locked up.
